### PR TITLE
Add CodeQL workflow

### PR DESCRIPTION
**Pull request by jorgectf**
Hello from [GitHub Security Lab](https://securitylab.github.com/)!

Your repository is critical to the security of the Open Source Software (OSS) ecosystem and as part of our mission to make OSS safer, we are contributing a [CodeQL configuration for code scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning-for-a-repository#setting-up-code-scanning-manually) to your repository. By enabling code scanning with CodeQL, you will be able to continuously analyze your code and surface potential vulnerabilities [before they can even reach your codebase](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/triaging-code-scanning-alerts-in-pull-requests#about-code-scanning-results-on-pull-requests). In fact, you may have seen some alerts already appearing on this pull request!

We’ve tested the configuration manually before opening this pull request and adjusted it to the needs of your particular repository, but feel free to tweak it further! Check [this page](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#editing-a-code-scanning-workflow) for detailed documentation.

Questions? Check out the FAQ below!

### FAQ
<details>
<summary>Click here to expand the FAQ section</summary>

#### How often will the code scanning analysis run?
By default, code scanning will trigger a scan with the CodeQL engine on the following events:
* On every pull request — to flag up potential security problems for you to investigate before merging a PR.
* On every push to your default branch and other protected branches — this keeps the analysis results on your repository’s *Security* tab up to date.
* Once a week at a fixed time — to make sure you benefit from the latest updated security analysis even when no code was committed or PRs were opened.

#### What will this cost?
Nothing! The CodeQL engine will run inside GitHub Actions, making use of your [unlimited free compute minutes for public repositories](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#about-billing-for-github-actions).

#### Where can I see the results of the analysis?
The results of the analysis will be available on the *Security* tab of your repository. You can find more information about the results [here](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/managing-code-scanning-alerts-for-your-repository#viewing-the-alerts-for-a-repository).

#### What types of problems does CodeQL find?
By default, code scanning runs the [`default` query suite](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/built-in-codeql-query-suites#default-query-suite).

#### How do I upgrade my CodeQL engine?
No need! New versions of the CodeQL analysis are constantly deployed on GitHub.com; your repository will automatically benefit from the most recently released version.

#### The analysis doesn’t seem to be working
If you get an error in GitHub Actions that indicates that CodeQL wasn’t able to analyze your code, please [follow the instructions here](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/troubleshooting-the-codeql-workflow) to debug the analysis.

#### Which source code hosting platforms does code scanning support?
GitHub code scanning is deeply integrated within GitHub itself. If you’d like to scan source code that is hosted elsewhere, we suggest that you create a mirror of that code on GitHub.

</details>

### CODE REVIEW

**Comment by droidmonkey** [comment]
Had to adjust our GitHub Actions security settings. Not sure how to re-run an action

**Comment by jorgectf** [comment]
> Had to adjust our GitHub Actions security settings. Not sure how to re-run an action

[Apparently](https://github.com/keepassxreboot/keepassxc/actions/runs/5376722246), there is a repository-level or organization-level setting that only allows `keepassxreboot`-owned Actions to run. Is that what you changed?

**Comment by droidmonkey** [comment]
Yes I allowed github managed actions as well

**Comment by jorgectf** [comment]
@droidmonkey: https://github.com/keepassxreboot/keepassxc/pull/9581/commits/ebc23218e4a7e13459824780a155f072cca986da should force the workflow to re-run.

**Comment by droidmonkey** [comment]
@jorgectf is it necessary to build the code base in order to perform a codeQL scan? Seems like it never runs any built binaries.

**Comment by jorgectf** [comment]
> @jorgectf is it necessary to build the code base in order to perform a codeQL scan? Seems like it never runs any built binaries.

Yes, CodeQL needs to trace the build to create the database for the analysis :)

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+        # For more details on CodeQL's query packs, refer to: https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
+        queries: security-and-quality
+
+    - if: matrix.language == 'cpp'
+      name: Build C++
+      run: |
+        mkdir build && cd build
+        cmake -DWITH_XC_ALL=ON -DWITH_TESTS=OFF ..
+        make -j 2
```
Comment (jorgectf) [inline .github/workflows/codeql.yml]:
```suggestion
        make -j $(nproc)
```

**Comment by droidmonkey** [comment]
Couple more changes, we shouldn't be analyzing qt generated code (moc_, ui_, etc) in the build folder. We should also exclude third-party code like zxcvbn.c and basestyle.cpp

**Pull request merged**